### ISSUE TITLE
16$: Vim: coc.nvim plugin

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#102]

**Comment by vegerot** [comment]
Idea: what if we made a Directly Integrated Completion plugin system for nvim?

**Comment by vegerot** [comment]
Okay okay okay.  I see your point.  Having it directly integrated might make it less accessible.  What if instead we worked on a Widely Available Plugin system for neovim?

**Comment by fannheyward** [comment]
I claim this for-hire task. I expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task.

https://github.com/fannheyward/coc-quick-lint-js

You can install with `Plug 'fannheyward/coc-quick-lint-js', {'do': 'yarn install --frozen-lockfile'}`

**Comment by fannheyward** [comment]
You don't need an extension/plugin for coc.nvim, just put this to your coc-settings.json:

```json
  "languageserver": {
    "quick-lint-js": {
      "command": "quick-lint-js",
      "args": ["--lsp"],
      "filetypes": [
        "javascript"
      ]
    }
  }
```

**Comment by vegerot** [comment]
@fannheyward  So perhaps all the “plugin” will do is bundle a copy of qljs and set this up

truth is, it should be nearly copy-paste of the VSCode extension

**Comment by fannheyward** [comment]
@vegerot there're two ways to make `quick-lint-js` work with VSCode https://github.com/quick-lint/quick-lint-js/tree/master/plugin:

- `vscode-lsp` works from `quick-lint-js --lsp`
- `vscode` bundled a Node.js N-API module

For the first way, I think it's unnecessary to write something like `coc-quick-lint-js` because we can use it from coc-settings.json with 5 lines of code.

For the second way, I've tried it but couldn't make it work with coc.nvim https://github.com/quick-lint/quick-lint-js/pull/330#issuecomment-960908253, coc.nvim provided similar API as VSCode, but there're still different. Maybe we can write something similar as https://github.com/quick-lint/quick-lint-js/blob/master/plugin/vscode/quick-lint-js-vscode-node.cpp but for coc.nvim.

**Comment by fannheyward** [comment]
In https://github.com/quick-lint/quick-lint-js/pull/330, I've tried three ways to make coc.nvim extension for quick-lint-js:

1. run `quick-lint-js` and parse the stdout/stderr to display
2. run `quick-lint-js-wasm` with Node.js, VSCode extension used this way too, and now it was deprecated
3. run `quick-lint-js-vscode-node.cpp` module, failed

We can build coc.nvim extension with `quick-lint-js --lsp`, but I think it's unnecessary  now.

**Comment by strager** [comment]
I implemented @fannheyward's suggestion in commit a755b0edf1b6e8550b558d0918eed185f6ffe802. Thanks for looking into this.
